File: PyProject/leetcode/history/symmetric-tree.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def is_symmetric(left: TreeNode, right: TreeNode) -> bool:
    if left is None and right is None:
        return True
    if left is None or right is None:
        return False

    if left.val == right.val:
        out = is_symmetric(left.left, right.right)
        inner = is_symmetric(left.right, right.left)
        return out and inner
    else:
        return False


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if root is None:
            return True
        else:
            return is_symmetric(root.left, root.right)

File: GoProject/leetcode/tree/symmetric-tree.go
package tree

func symmetry(nodes []*TreeNode) bool {
	res := []int{}
	rets := []*TreeNode{}

	if len(nodes)%2 != 0 && len(nodes) != 1 {
		return false
	}
	for _, node := range nodes {
		res = append(res, node.Val)
		if node != nil {
			if node.Left != nil {
				rets = append(rets, node.Left)
			}
			if node.Right != nil {
				rets = append(rets, node.Right)
			}
		}
	}

	if len(res)%2 != 0 && len(res) != 1 {
		return false
	}

	for i := 0; i < len(res)/2; i++ {
		if res[i] != res[len(res)-1-i] {
			return false
		}
	}
	if len(rets) == 0 {
		return true
	}

	return symmetry(rets)
}

func isSymmetric(root *TreeNode) bool {
	return symmetry([]*TreeNode{root})
}
